### PR TITLE
make an ocsp request

### PR DESCRIPTION
**Pull request by reaperhulk**
WIP, let's discuss.

### CODE REVIEW

File: tests/x509/test_ocsp.py [hunk 1]
```diff
@@ -113,3 +115,20 @@ def test_invalid_serialize_encoding(self):
...
+        x509.load_pem_x509_certificate,
+        backend
+    )
+    issuer = _load_cert(
+        os.path.join("x509", "rapidssl_sha256_ca_g3.pem"),
+        x509.load_pem_x509_certificate,
+        backend
+    )
+    req = ocsp.create_ocsp_request_from_cert(cert, issuer, hashes.SHA1())
+    with open("/Users/pkehrer/Desktop/out.der", "wb") as f:
```
Comment (alex) [inline tests/x509/test_ocsp.py]: this is a cool unit test

Comment (reaperhulk) [reply tests/x509/test_ocsp.py]: Yep :) I have it doing something different locally now but didn't see much point to pushing it when I know we'll be going through other revisions

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 2]
```diff
@@ -1430,6 +1430,22 @@ def load_der_ocsp_request(self, data):
         request = self._ffi.gc(request, self._lib.OCSP_REQUEST_free)
         return _OCSPRequest(self, request)
 
+    def create_ocsp_request(self, requests):
+        ocsp_req = self._lib.OCSP_REQUEST_new()
+        self.openssl_assert(ocsp_req != self._ffi.NULL)
+        ocsp_req = self._ffi.gc(ocsp_req, self._lib.OCSP_REQUEST_free)
+        for request in requests:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: tuple unpacking

File: src/cryptography/x509/ocsp.py [hunk 3]
```diff
@@ -25,6 +25,21 @@ def load_der_ocsp_request(data):
     return backend.load_der_ocsp_request(data)
 
 
+class OCSPRequestBuilder(object):
+    def __init__(self, requests=[]):
+        self._requests = requests
+
+    def add_request(self, cert, issuer, algorithm):
+        return OCSPRequestBuilder(self._requests + [(cert, issuer, algorithm)])
```
Comment (alex) [inline src/cryptography/x509/ocsp.py]: need type checking here

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 4]
```diff
@@ -1430,6 +1430,22 @@ def load_der_ocsp_request(self, data):
         request = self._ffi.gc(request, self._lib.OCSP_REQUEST_free)
         return _OCSPRequest(self, request)
 
+    def create_ocsp_request(self, requests):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: this should take the builder I think

File: src/cryptography/x509/ocsp.py [hunk 5]
```diff
@@ -25,6 +25,29 @@ def load_der_ocsp_request(data):
...
+    def add_request(self, cert, issuer, algorithm):
+        allowed_hashes = (
+            hashes.SHA1, hashes.SHA224, hashes.SHA256,
+            hashes.SHA384, hashes.SHA512
+        )
+        if not isinstance(algorithm, allowed_hashes):
+            raise ValueError(
+                "Algorithm must be SHA1, SHA224, SHA256, SHA384, or SHA512"
+            )
+        return OCSPRequestBuilder(self._requests + [(cert, issuer, algorithm)])
```
Comment (alex) [inline src/cryptography/x509/ocsp.py]: Add type checks for `cert` and `issuer`

**Pull request merged**